dali/pipeline/data/tensor_vector.cc

```diff
@@ -172,240 +206,299 @@ void CopyImpl(DstBatch<DstBackend> &dst, const SrcBatch<SrcBackend> &src, const
+                              shape()[sample_idx], src.shape(),
+                              make_string(" for ", sample_idx, "."));
+
+  shape_.set_tensor_shape(sample_idx, src.shape());
+  tensors_[sample_idx].Copy(src, order);
+  if (src.GetLayout().empty() && !GetLayout().empty()) {
+    tensors_[sample_idx].SetLayout(GetLayout());
+  }
 }
 
+
 template <typename Backend>
 void TensorVector<Backend>::set_sample_dim(int sample_dim) {
   DALI_ENFORCE(
       !has_data(),
       "Setting sample dim is not allowed when batch is already allocated, use Resize instead.");
   sample_dim_ = sample_dim;
+  shape_.resize(shape_.num_samples(), sample_dim);
 }
 
+
 template <typename Backend>
 size_t TensorVector<Backend>::nbytes() const noexcept {
-  if (state_ == State::contiguous) {
-    return tl_->nbytes();
+  if (IsContiguous()) {
+    return contiguous_buffer_.nbytes();
   }
   // else
   size_t nbytes = 0;
   for (const auto &t : tensors_) {
-    nbytes += t->nbytes();
+    nbytes += t.nbytes();
   }
   return nbytes;
 }
 
 
 template <typename Backend>
 size_t TensorVector<Backend>::capacity() const noexcept {
-  if (state_ == State::contiguous) {
-    return tl_->capacity();
+  if (IsContiguous()) {
+    return contiguous_buffer_.capacity();
   }
   // else
   size_t capacity = 0;
   for (const auto &t : tensors_) {
-    capacity += t->capacity();
+    capacity += t.capacity();
   }
   return capacity;
 }
 
+
 template <typename Backend>
 std::vector<size_t> TensorVector<Backend>::_chunks_nbytes() const {
-  if (state_ == State::contiguous) {
-    return {tl_->nbytes()};
+  if (IsContiguous()) {
+    return {contiguous_buffer_.nbytes()};
   }
   // else
   std::vector<size_t> result(tensors_.size());
   for (size_t i = 0; i < tensors_.size(); i++) {
-    result[i] = tensors_[i]->nbytes();
+    result[i] = tensors_[i].nbytes();
   }
   return result;
 }
 
 
 template <typename Backend>
 std::vector<size_t> TensorVector<Backend>::_chunks_capacity() const {
-  if (state_ == State::contiguous) {
-    return {tl_->capacity()};
+  if (IsContiguous()) {
+    return {contiguous_buffer_.capacity()};
   }
   // else
   std::vector<size_t> result(tensors_.size());
   for (size_t i = 0; i < tensors_.size(); i++) {
-    result[i] = tensors_[i]->capacity();
+    result[i] = tensors_[i].capacity();
   }
   return result;
 }
 
 
 template <typename Backend>
-TensorListShape<> TensorVector<Backend>::shape() const {
-  if (state_ == State::contiguous) {
-    return tl_->shape();
-  }
-  if (curr_num_tensors_ == 0) {
-    return {};
-  }
-  TensorListShape<> result(curr_num_tensors_, tensors_[0]->ndim());
-  for (int i = 0; i < curr_num_tensors_; i++) {
-    result.set_tensor_shape(i, tensors_[i]->shape());
-  }
-  return result;
+const TensorListShape<> &TensorVector<Backend>::shape() const & {
+  return shape_;
 }
 
+
 template <typename Backend>
 void TensorVector<Backend>::set_order(AccessOrder order, bool synchronize) {
```

Comment:
Does it? It only keeps the order updated there in case we need it there, rather than update it late when it's needed.
In that case the old version maps to the one after refactor, I think. Do you expect any problem with that approach?

Reply:
not really, it just looked odd at first that we are manipulating `contiguous_buffer_` unconditionally. But after your explanation is clear now